src/certification.jl

```diff
@@ -0,0 +1,282 @@
+    duplicates = find_duplicates(certified)
+    CertificationResult(certified, duplicates)
+end
+function _certify(F::AbstractSystem, r::PathResult, cache::CertifyCache, index = nothing)
+    _certify(F, solution(r), cache, index)
+end
+function _certify(
+    F::AbstractSystem,
+    x̃::Vector{ComplexF64},
+    cache::CertifyCache,
+    index = nothing,
+)
+    res = newton(
+        F,
+        x̃,
+        InfNorm(),
+        cache.newton_cache;
+        abs_tol = 4 * eps(),
+        rel_tol = 4 * eps(),
+        extended_precision = true,
+    )
+    x̃ = res.x
+    acc = res.accuracy
+    # heuristic, try different distances
+    for p in [0.5, 0.75, 0.95]
+        x₁, x₀ = krawczyc_step_ε_inflation(F, x̃, acc^p, cache)
+        if all2(⪽, x₁, x₀)
+            # check whether it is real
+            # 1) check whether uniqueness region contains a real point
+            if all(xi -> 0 ∈ imag(xi), x₀)
+                # 2) project x̃ onto real part
+                ỹ = real.(x̃)
+                y₁, y₀ = krawczyc_step_ε_inflation(F, ỹ, acc^p, cache)
+                if all2(⪽, y₁, y₀)
+                    return CertifiedSolution(
+                        solution = x̃,
+                        uniqueness_region = x₀,
+                        real_solution = ỹ,
+                        real_uniqueness_region = y₀,
+                        index = index,
+                    )
+                end
+            end
+            return CertifiedSolution(solution = x̃, uniqueness_region = x₀, index = index)
+        end
+    end
+    nothing
+end
+
+function jacobian(F::AbstractSystem, x::AbstractVector{IComplexF64}, p = nothing)
+    u, U = zeros(IComplexF64, size(F, 1)), zeros(IComplexF64, size(F))
+    evaluate_and_jacobian!(u, U, F, x, p)
+    U
+end
+
+# complex case
+function krawczyc_step_ε_inflation(
+    F,
+    x̃::AbstractVector{<:Complex},
+    ε::Real,
+    cache::CertifyCache,
+)
+    x = @. x̃ + complex(Interval(-ε, ε), Interval(-ε, ε))
+    krawczyc_step(F, x, x̃, cache), x
+end
+function krawczyc_step(
+    F::AbstractSystem,
+    x,
+    x̃::AbstractVector{<:Complex},
+    cache::CertifyCache,
+)
+    # TODO: use of LU fac gives catastrophic errors in the ldiv step
+    # probably since the division algorithm for IComplex fucks something up
+    @unpack r, J, Δx, x_ext = cache.newton_cache
+    @unpack C, Ir, IJ = cache
+    # Compute r = F(x̃) and J = J_F(x̃)
+    evaluate_and_jacobian!(r, matrix(J), F, x̃)
+    updated!(J)
+    x_ext .= x̃
+    evaluate!(r, F, x_ext)
+
+    # we only need J_F(x), however, an abstract system only needs to implement
+    # evaluate_and_jacobian! currently. So we use this
+    evaluate_and_jacobian!(Ir, IJ, F, x)
+
+    # TODO: Use our fast linear algebra, make non-allocating
+    LA.inv!(C, J)
+    # TODO: make non-allocating, we currently spend 80+ % our time here!!!
+    x̃ - C * r + (LinearAlgebra.I - C * IJ) * (x - x̃)
```

Comment:
This line is where we currently spent the vast majority of our time. So we definitely need to optimize this

Reply:
Step 1 is probably rewriting this:
```
  x̃ - C * r + (LinearAlgebra.I - C * IJ) * (x - x̃)
= x̃ - C * r + (x - x̃) - C * IJ * (x - x̃)
= x - C * r - C * IJ * (x - x̃)
= x - C * (r + IJ * (x - x̃))
```
(we just need to make sure that this doesn't introduce more FP error)

Reply:
Also the diameter in `LinearAlgebra.I - C * IJ` is approximately `cond(J) *eps()`.
From this follows that the radius in the epsilon inflation has to be less than `cond(J) *eps()`.

Reply:
The rewriting doesn't work since `IJ * (x - x̃)` widens the interval by too much